# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_id_text=[]
    for s in source_text.split("\n"):
        source_id_text.append([source_vocab_to_int[w] for w in s.split()])
    
    target_id_text=[]
    for t in target_text.split("\n"):
        target_id_text.append([target_vocab_to_int[w] for w in t.split()] + [target_vocab_to_int['<EOS>']])
        
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


E:\tom\Anaconda3\envs\tf-1-1\lib\site-packages\ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputp = tf.placeholder(tf.int32,[None,None], name='input')
 
    targets = tf.placeholder(tf.int32,[None,None], name='targets')
    
    leaning_rate = tf.placeholder(tf.float32, name='learning_rate')
    
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    
    target_sequence_length = tf.placeholder(tf.int32, [None], name='target_sequence_length')
    
    max_target_len = tf.reduce_max(target_sequence_length, name='max_target_len')
    
    source_sequence_length = tf.placeholder(tf.int32, [None], name='source_sequence_length')
    
    return inputp, targets, leaning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    #Get rid of last character - not needed as input to next time-slice in decoder
    #https://discussions.udacity.com/t/what-does-this-function-do-exactlyending-tf-strided-slice-targets-0-0-batch-size-1-1-1/240611/2
    go_added = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), target_data], 1)
    last_removed = tf.strided_slice(go_added, [0, 0], [batch_size, -1], [1, 1])

    return last_removed

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    
    encoded_i = tf.contrib.layers.embed_sequence(rnn_inputs, embed_dim=encoding_embedding_size, 
                                               vocab_size=source_vocab_size)
    
    def build_cell(num_units, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        
        return drop
    
    # Stack up multiple LSTM layers, for deep learning
    enc_cell = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    #initial_state = enc_cell.zero_state(rnn_inputs.get_shape().as_list()[0] , tf.float32)
    
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, encoded_i, sequence_length=source_sequence_length, dtype=tf.float32)

    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """


    # Helper for the training process. Used by BasicDecoder to read inputs.
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                                        sequence_length=target_sequence_length,
                                                        time_major=False)
        
        
    # Basic decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                           training_helper,
                                                           encoder_state,
                                                           output_layer) 
        
    # Perform dynamic decoding using the decoder
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder,
                                                                impute_finished=True,
                                                                maximum_iterations=max_summary_length)[0]
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function

    # Helper for the training process. Used by BasicDecoder to read inputs.
    helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(embedding=dec_embeddings,
                                                      end_token=end_of_sequence_id,
                                                      start_tokens=tf.fill([batch_size], 
                                                                           start_of_sequence_id))
        
        
    # Basic decoder
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                  helper,
                                                  encoder_state,
                                                  output_layer) 
        
    # Perform dynamic decoding using the decoder
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(decoder,
                                                                impute_finished=True,
                                                                maximum_iterations=
                                                                max_target_sequence_length)[0]
    return training_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    embedding = tf.Variable(tf.truncated_normal((target_vocab_size, decoding_embedding_size),stddev=0.1))
    
    embed_input = tf.nn.embedding_lookup(embedding, dec_input) 
 
    def build_cell(num_units, keep_prob):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return drop
    
    # Stack up multiple LSTM layers, for deep learning
    dec_cell = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    #initial_state = enc_cell.zero_state(rnn_inputs.get_shape().as_list()[0] , tf.float32)
    
    output_layer = Dense (target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("dec", reuse=False):
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, embed_input, 
                                                       target_sequence_length, 
                                                        max_target_sequence_length, output_layer, 
                                                       keep_prob)
    
    with tf.variable_scope("dec", reuse=True):
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, embedding, 
                                                        target_vocab_to_int['<GO>'], 
                                                        target_vocab_to_int['<EOS>'],
                                                        max_target_sequence_length, 
                                                        target_vocab_size, output_layer, 
                                                        batch_size, keep_prob) 
        
    return training_decoder_output, inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    _, enc_state = encoding_layer(input_data, rnn_size, num_layers, 
                               keep_prob,  source_sequence_length, 
                               source_vocab_size, enc_embedding_size)
    
    target_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)

    t_out, inf_out = decoding_layer(target_input, enc_state, target_sequence_length, 
                                    max_target_sentence_length, rnn_size, num_layers, 
                                    target_vocab_to_int, target_vocab_size, batch_size, 
                                    keep_prob, dec_embedding_size)    
    
    return t_out, inf_out


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 5
# Batch Size
batch_size = 120
# RNN Size
rnn_size = 300
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = .0004
# Dropout Keep Probability
keep_probability = .5
display_step = 10



### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   10/1148 - Train Accuracy: 0.2821, Validation Accuracy: 0.3402, Loss: 4.2704
Epoch   0 Batch   20/1148 - Train Accuracy: 0.3108, Validation Accuracy: 0.3648, Loss: 3.4452
Epoch   0 Batch   30/1148 - Train Accuracy: 0.3417, Validation Accuracy: 0.3973, Loss: 3.0706
Epoch   0 Batch   40/1148 - Train Accuracy: 0.3421, Validation Accuracy: 0.4098, Loss: 2.9114
Epoch   0 Batch   50/1148 - Train Accuracy: 0.3983, Validation Accuracy: 0.4553, Loss: 2.6960
Epoch   0 Batch   60/1148 - Train Accuracy: 0.4075, Validation Accuracy: 0.4686, Loss: 2.5594
Epoch   0 Batch   70/1148 - Train Accuracy: 0.3858, Validation Accuracy: 0.4602, Loss: 2.5147
Epoch   0 Batch   80/1148 - Train Accuracy: 0.4188, Validation Accuracy: 0.4602, Loss: 2.3273
Epoch   0 Batch   90/1148 - Train Accuracy: 0.4454, Validation Accuracy: 0.5064, Loss: 2.2856
Epoch   0 Batch  100/1148 - Train Accuracy: 0.4404, Validation Accuracy: 0.4864, Loss: 2.1492
Epoch   0 Batch  110/1148 - Train Accuracy: 0.3996, Validati

Epoch   0 Batch  890/1148 - Train Accuracy: 0.7398, Validation Accuracy: 0.7405, Loss: 0.5774
Epoch   0 Batch  900/1148 - Train Accuracy: 0.7052, Validation Accuracy: 0.7303, Loss: 0.5781
Epoch   0 Batch  910/1148 - Train Accuracy: 0.7104, Validation Accuracy: 0.7174, Loss: 0.5627
Epoch   0 Batch  920/1148 - Train Accuracy: 0.7308, Validation Accuracy: 0.7337, Loss: 0.5491
Epoch   0 Batch  930/1148 - Train Accuracy: 0.7483, Validation Accuracy: 0.7371, Loss: 0.5785
Epoch   0 Batch  940/1148 - Train Accuracy: 0.7117, Validation Accuracy: 0.7553, Loss: 0.5960
Epoch   0 Batch  950/1148 - Train Accuracy: 0.7508, Validation Accuracy: 0.7292, Loss: 0.5755
Epoch   0 Batch  960/1148 - Train Accuracy: 0.7063, Validation Accuracy: 0.7189, Loss: 0.6194
Epoch   0 Batch  970/1148 - Train Accuracy: 0.7079, Validation Accuracy: 0.7163, Loss: 0.5287
Epoch   0 Batch  980/1148 - Train Accuracy: 0.7009, Validation Accuracy: 0.7413, Loss: 0.5715
Epoch   0 Batch  990/1148 - Train Accuracy: 0.7137, Validati

Epoch   1 Batch  630/1148 - Train Accuracy: 0.8548, Validation Accuracy: 0.8621, Loss: 0.1904
Epoch   1 Batch  640/1148 - Train Accuracy: 0.8771, Validation Accuracy: 0.8648, Loss: 0.2132
Epoch   1 Batch  650/1148 - Train Accuracy: 0.8600, Validation Accuracy: 0.8727, Loss: 0.2050
Epoch   1 Batch  660/1148 - Train Accuracy: 0.8904, Validation Accuracy: 0.8792, Loss: 0.1836
Epoch   1 Batch  670/1148 - Train Accuracy: 0.8800, Validation Accuracy: 0.8780, Loss: 0.2188
Epoch   1 Batch  680/1148 - Train Accuracy: 0.8888, Validation Accuracy: 0.8723, Loss: 0.1776
Epoch   1 Batch  690/1148 - Train Accuracy: 0.8588, Validation Accuracy: 0.8913, Loss: 0.1915
Epoch   1 Batch  700/1148 - Train Accuracy: 0.8796, Validation Accuracy: 0.8629, Loss: 0.1871
Epoch   1 Batch  710/1148 - Train Accuracy: 0.8667, Validation Accuracy: 0.8777, Loss: 0.1795
Epoch   1 Batch  720/1148 - Train Accuracy: 0.8738, Validation Accuracy: 0.8803, Loss: 0.2093
Epoch   1 Batch  730/1148 - Train Accuracy: 0.8712, Validati

Epoch   2 Batch  370/1148 - Train Accuracy: 0.9274, Validation Accuracy: 0.8939, Loss: 0.0967
Epoch   2 Batch  380/1148 - Train Accuracy: 0.8946, Validation Accuracy: 0.8864, Loss: 0.1217
Epoch   2 Batch  390/1148 - Train Accuracy: 0.9500, Validation Accuracy: 0.8898, Loss: 0.0902
Epoch   2 Batch  400/1148 - Train Accuracy: 0.9371, Validation Accuracy: 0.8890, Loss: 0.1193
Epoch   2 Batch  410/1148 - Train Accuracy: 0.9237, Validation Accuracy: 0.9000, Loss: 0.0931
Epoch   2 Batch  420/1148 - Train Accuracy: 0.9067, Validation Accuracy: 0.8894, Loss: 0.1053
Epoch   2 Batch  430/1148 - Train Accuracy: 0.8583, Validation Accuracy: 0.9072, Loss: 0.1179
Epoch   2 Batch  440/1148 - Train Accuracy: 0.9133, Validation Accuracy: 0.8693, Loss: 0.0858
Epoch   2 Batch  450/1148 - Train Accuracy: 0.9096, Validation Accuracy: 0.9102, Loss: 0.0994
Epoch   2 Batch  460/1148 - Train Accuracy: 0.8925, Validation Accuracy: 0.9053, Loss: 0.0852
Epoch   2 Batch  470/1148 - Train Accuracy: 0.9219, Validati

Epoch   3 Batch  110/1148 - Train Accuracy: 0.9127, Validation Accuracy: 0.9152, Loss: 0.0686
Epoch   3 Batch  120/1148 - Train Accuracy: 0.9167, Validation Accuracy: 0.9178, Loss: 0.0746
Epoch   3 Batch  130/1148 - Train Accuracy: 0.9308, Validation Accuracy: 0.9170, Loss: 0.0661
Epoch   3 Batch  140/1148 - Train Accuracy: 0.9129, Validation Accuracy: 0.9030, Loss: 0.0803
Epoch   3 Batch  150/1148 - Train Accuracy: 0.9600, Validation Accuracy: 0.9011, Loss: 0.0585
Epoch   3 Batch  160/1148 - Train Accuracy: 0.9349, Validation Accuracy: 0.8947, Loss: 0.0721
Epoch   3 Batch  170/1148 - Train Accuracy: 0.9417, Validation Accuracy: 0.9288, Loss: 0.0583
Epoch   3 Batch  180/1148 - Train Accuracy: 0.9175, Validation Accuracy: 0.9045, Loss: 0.1012
Epoch   3 Batch  190/1148 - Train Accuracy: 0.9225, Validation Accuracy: 0.9178, Loss: 0.0709
Epoch   3 Batch  200/1148 - Train Accuracy: 0.9350, Validation Accuracy: 0.9129, Loss: 0.0656
Epoch   3 Batch  210/1148 - Train Accuracy: 0.9567, Validati

Epoch   3 Batch  990/1148 - Train Accuracy: 0.9113, Validation Accuracy: 0.9057, Loss: 0.0496
Epoch   3 Batch 1000/1148 - Train Accuracy: 0.9492, Validation Accuracy: 0.9273, Loss: 0.0640
Epoch   3 Batch 1010/1148 - Train Accuracy: 0.9377, Validation Accuracy: 0.9311, Loss: 0.0601
Epoch   3 Batch 1020/1148 - Train Accuracy: 0.9458, Validation Accuracy: 0.9258, Loss: 0.0472
Epoch   3 Batch 1030/1148 - Train Accuracy: 0.9203, Validation Accuracy: 0.9405, Loss: 0.0644
Epoch   3 Batch 1040/1148 - Train Accuracy: 0.9500, Validation Accuracy: 0.9284, Loss: 0.0572
Epoch   3 Batch 1050/1148 - Train Accuracy: 0.9029, Validation Accuracy: 0.9337, Loss: 0.0568
Epoch   3 Batch 1060/1148 - Train Accuracy: 0.9392, Validation Accuracy: 0.9303, Loss: 0.0579
Epoch   3 Batch 1070/1148 - Train Accuracy: 0.9487, Validation Accuracy: 0.9383, Loss: 0.0617
Epoch   3 Batch 1080/1148 - Train Accuracy: 0.9679, Validation Accuracy: 0.9424, Loss: 0.0365
Epoch   3 Batch 1090/1148 - Train Accuracy: 0.9643, Validati

Epoch   4 Batch  730/1148 - Train Accuracy: 0.9400, Validation Accuracy: 0.9511, Loss: 0.0408
Epoch   4 Batch  740/1148 - Train Accuracy: 0.9595, Validation Accuracy: 0.9530, Loss: 0.0497
Epoch   4 Batch  750/1148 - Train Accuracy: 0.9604, Validation Accuracy: 0.9557, Loss: 0.0521
Epoch   4 Batch  760/1148 - Train Accuracy: 0.9587, Validation Accuracy: 0.9564, Loss: 0.0410
Epoch   4 Batch  770/1148 - Train Accuracy: 0.9471, Validation Accuracy: 0.9595, Loss: 0.0474
Epoch   4 Batch  780/1148 - Train Accuracy: 0.9615, Validation Accuracy: 0.9621, Loss: 0.0441
Epoch   4 Batch  790/1148 - Train Accuracy: 0.9417, Validation Accuracy: 0.9580, Loss: 0.0560
Epoch   4 Batch  800/1148 - Train Accuracy: 0.9350, Validation Accuracy: 0.9515, Loss: 0.0473
Epoch   4 Batch  810/1148 - Train Accuracy: 0.9276, Validation Accuracy: 0.9402, Loss: 0.0528
Epoch   4 Batch  820/1148 - Train Accuracy: 0.9408, Validation Accuracy: 0.9530, Loss: 0.0484
Epoch   4 Batch  830/1148 - Train Accuracy: 0.9313, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [23]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    return [vocab_to_int[i] if i in vocab_to_int else vocab_to_int['<UNK>'] for i in sentence.split()]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [33]:
translate_sentence = 'he saw a old yellow truck .'
translate_sentence = "new jersey is sometimes warm during autumn , and it is snowy and cold in winter ."


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [135, 9, 98, 92, 88, 129, 27, 190, 96, 63, 98, 34, 96, 45, 226, 6, 124]
  English Words: ['new', 'jersey', 'is', 'sometimes', 'warm', 'during', 'autumn', ',', 'and', 'it', 'is', 'snowy', 'and', 'cold', 'in', 'winter', '.']

Prediction
  Word Ids:      [224, 5, 291, 352, 149, 267, 357, 58, 35, 251, 239, 87, 266, 144, 76, 311, 51, 1]
  French Words: new jersey est parfois chaud au cours de l' automne , et il gèle en hiver . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.